### PR TITLE
Add secrets-backends doc on using Vault+Astro

### PR DESCRIPTION
**Pull request by petedejoy**

### CODE REVIEW

File: v0.15/secrets-backends.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
...rflow has long been constrained to Airflow's native [connections](https://airflow.apache.org/docs/stable/howto/connection/index.html) and [variables](https://airflow.apache.org/docs/stable/concepts.html#variables) features. The default connections & variables features are great for getting started, but they don't scale particularly well to multiple Airflow environments or a more complex cloud automation setup that needs to leverage replication of Airflow environments; using only those mechanisms for adding connections, every time a team needs to spin up a new Airflow, they need to manually re-input all of their connection information directly via the Airflow UI. As of Airflow 1.10.10, this limitation is no longer a concern; users can [manage connections from an external secrets backend](https://airflow.apache.org/docs/1.10.10/howto/use-alternative-secrets-backend.html) such as [Hashicorp Vault](https://www.vaultproject.io/), [AWS SSM](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html), or [GCP Secret Manager](https://www.google.com/search?q=gcp+secrets+manager&oq=gcp+secrets+manager&aqs=chrome.0.0l8.2480j0j4&sourceid=chrome&ie=UTF-8).
```
Comment (vparekh94) [inline v0.15/secrets-backends.md:8]: We don't need all this stuff to start - we should just dive into "Airflow 1.10.10... allows you to"

File: v0.15/secrets-backends.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    ====== Data ======
+    Key         Value
+    ---         -----
+    conn_uri    smtps://user:host@relay.example.com:465
+
+### Configure Your Dockerfile
+
+Now that we have a Vault server running and a secret injected there, let's configure our Astronomer project's Dockerfile to use this Vault server as the default secrets backend for your Airflow instance. To set this configuration up, we'll need to do a few things:
+
+1. Create a `.env` file in your local project directory and add the following environment variables to be used in your _local_ Airflow instance:
```
Comment (vparekh94) [inline v0.15/secrets-backends.md:52]: Isn't a `.env` file created by default in by the CLI?

File: v0.15/secrets-backends.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+Once you've added this DAG to your project, run `astro dev start` to spin up your project locally. Trigger this new test DAG and confirm that your `smtp_default` connection information is being printed in the logs of the task. Note that, if you've written other secrets to your Vault server's `/connections` path, you should be able to test those here as well just by changing the `my_conn_id` value.
+
+This works because we set the `connections_path` in our `AIRFLOW__SECRETS__BACKEND_KWARGS` to be `connections`. You are welcome to change this path name if you'd prefer to access variables from a different Vault directory.
+
+### Use Vault Secrets in your DAGs
+
+Once you've established that your connection is working, you can leverage the `BaseHook.get_connection(kwargs['conn_id'])` method to use your connections in your own DAG code. You can also pass the `conn_id` of your secret directly to most operators, which will take care of instantiating the connection using a pre-built hook.
+
+### Deploy to Production
```
Comment (vparekh94) [inline v0.15/secrets-backends.md:99]: Deploy to Astronomer

File: v0.15/secrets-backends.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+Once you've confirmed that your connections are being imported correctly locally, head to the Astronomer UI to set your production environment's environment variables. In the `Deployment>Configure>Environment Variables` section of our UI, set your `VAULT__ROOT_TOKEN` and `VAULT__URL`.
+
+Once your env vars are set, deploy your local project up to Astronomer via `astro deploy`. You now should be able to observe your connection information being pulled from Vault on Astronomer!
+
+### Pulling Variables from Vault
```
Comment (vparekh94) [inline v0.15/secrets-backends.md:105]: We should add a screenshot of the logs of a task that's pulling the creds of a  task here.

File: v0.15/secrets-backends.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+2. Add the following lines to your `Dockerfile` to instantiate those vault-specific environment variables and use them in the context of Airflow's `AIRFLOW__SECRETS__BACKEND` configuration:
+
+        ENV VAULT__ROOT_TOKEN $VAULT__ROOT_TOKEN
+        ENV VAULT__URL $VAULT__URL
+        ENV AIRFLOW__SECRETS__BACKEND="airflow.contrib.secrets.hashicorp_vault.VaultBackend"
+        ENV AIRFLOW__SECRETS__BACKEND_KWARGS='{"url":$VAULT__URL,"token": $VAULT__ROOT_TOKEN,"connections_path": "connections","variables_path": "variables","kv_engine_version":1}'
+
+   This tells Airflow to look for connection information at the `secret/connections/*` path in your Vault server.
+
+### Test The Connection
```
Comment (vparekh94) [inline v0.15/secrets-backends.md:68]: I don't think we should show a user how to put those creds from vault into plaintext. If a user wants to do that, they can figure it out themselves. We should just show an example on using a connection for a task

File: v0.15/secrets-backends.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+### Configure Your Dockerfile
+
+Now that we have a Vault server running and a secret injected there, let's configure our Astronomer project's Dockerfile to use this Vault server as the default secrets backend for your Airflow instance. To set this configuration up, we'll need to do a few things:
+
+1. Create a `.env` file in your local project directory and add the following environment variables to be used in your _local_ Airflow instance:
+
+        VAULT__ROOT_TOKEN="YOUR-ROOT-TOKEN"
+        VAULT__URL="YOUR-VAULT-URL"
+
+   This will allow us to keep our connection to the Vault server secure; be sure to keep you `.env` in your `.gitignore` when you deploy these changes to your project; we'll set these variables separately in production.
```
Comment (vparekh94) [inline v0.15/secrets-backends.md:57]: Our cli defaults to creating a `.gitignore` that has the `.env` in there

**Comment by petedejoy** [comment]
@kaxil no rush on this, but can you check out my SSM Parameter Store section and leave me some feedback when you get a chance? Less confident about that one since I wasn't able to actually test it in the same as as Vault.

File: v0.15/secrets-backends.md (hunk 5 above)
Comment (petedejoy) [reply v0.15/secrets-backends.md]: I don't know, I think we need to include this section. Even knowing these steps, _so_ much can go wrong (ie. misformatted connection string, vault gets sealed, invalid root token, etc etc) and this DAG is designed to give some confirmation that things work with some logs that make sense.

File: v0.15/secrets-backends.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    version          1
+
+    ====== Data ======
+    Key         Value
+    ---         -----
+    conn_uri    smtps://user:host@relay.example.com:465
+
+### Configure Your Dockerfile
+
+Now that we have a Vault server running and a secret injected there, let's configure our Astronomer project's Dockerfile to use this Vault server as the default secrets backend for your Airflow instance. To set this configuration up, we'll need to do a few things:
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:46]: Are these Env Vars? Not a big deal on this one, but I'd rephrase:

> Now that you have a secret injected in your Vault server, let's set a few Environment Variables via your `Dockerfile` on Astronomer to allow Airflow to pull from that Vault server.
> 
> To set up that configuration, follow the steps below.

File: v0.15/secrets-backends.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+1. Add the following environment variables to the `.env` file in your project directory to be used in your _local_ Airflow instance:
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:48]: `environment variables` --> `Environment Variables`

File: v0.15/secrets-backends.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+        VAULT__ROOT_TOKEN="YOUR-ROOT-TOKEN"
+        VAULT__URL="YOUR-VAULT-URL"
+
+   This will allow us to keep our connection to the Vault server secure; be sure to keep you `.env` in your `.gitignore` when you deploy these changes to your project; we'll set these variables separately in production.
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:53]: Wording rec:

> **Note:** Be sure to keep the `.env` file in your `.gitignore` when you deploy these changes to your project to prevent the tokens you set here from being exposed in plain-text to your upstream code. We'll set these variables separately in production.

File: v0.15/secrets-backends.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    deletion_time    n/a
+    destroyed        false
+    version          1
+
+    ====== Data ======
+    Key         Value
+    ---         -----
+    conn_uri    smtps://user:host@relay.example.com:465
+
+### Configure Your Dockerfile
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:44]: After reading this I realized this process is local-only before you move to setting this up and testing it on an Airflow Deployment on Astronomer - is it worth calling that out somewhere, whether it's in the title of this section or as intro context below it?

File: v0.15/secrets-backends.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    with DAG('example_secrets_dags', start_date=datetime(2020, 1, 1), schedule_interval=None) as dag:
+    
+    
+        test_task = PythonOperator(
+            task_id='test-task',
+            python_callable=get_secrets,
+            op_kwargs={'my_conn_id': 'smtp_default'},
+     )
+
+Once you've added this DAG to your project, run `astro dev start` to spin up your project locally. Trigger this new test DAG and confirm that your `smtp_default` connection information is being printed in the logs of the task. Note that, if you've written other secrets to your Vault server's `/connections` path, you should be able to test those here as well just by changing the `my_conn_id` value.
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:91]: I'd break this up into a step-by-step list for easy reading:

> 1. Run `astro dev start` to instantiate Airflow locally
> 2. Trigger your new test DAG via the Airflow UI at http://localhost:8080/admin/
> 3. Click on `test-task` > `View Logs`
> 4. Confirm your `smtp_default` connection information is being printed in the logs of the task
> 
> > **Note:** If you've written other secrets to your Vault server's `/connections` path, you should be able to test those here as well by changing the `my_conn_id` value.

File: v0.15/secrets-backends.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+### Deploy to Astronomer
+
+Once you've confirmed that your connections are being imported correctly locally, head to the Astronomer UI to set your production environment's environment variables. In the `Deployment>Configure>Environment Variables` section of our UI, set your `VAULT__ROOT_TOKEN` and `VAULT__URL`.
+
+Once your env vars are set, deploy your local project up to Astronomer via `astro deploy`. You now should be able to observe your connection information being pulled from Vault on Astronomer!
+
+### Pulling Variables from Vault
+
+Note that, while the above section required that you add your connection information as a conn_uri to Vault, you can also pull Airflow Variables from Vault. To accomplish this, you'll need to add your variable to vault via the following syntax:
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:107]: Nice on including instructions for managing Variables - might add something about that at the very top of the doc?

File: v0.15/secrets-backends.md [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+    vault kv put airflow/variables/hello value=world
+
+This works because we set the `variables_path` in our `AIRFLOW__SECRETS__BACKEND_KWARGS` to be `variables`. You are welcome to change this path name if you'd prefer to access variables from a different Vault directory.
+
+Now, if you print `Variables.get('hello')` in Airflow, it will print the value of that variable; in our case, that's `world`.
+
+## AWS SSM Parameter Store
+
+In this section, we'll walk through how to use AWS SSM Parameter Store as a secrets backend for your Airflow environment running on Astronomer. For the purposes of this walkthrough, we'll assume you already have a SSM Parameter Store instance running.
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:117]: Same thing on pre-rec section, sounds like it'd make sense here too

File: v0.15/secrets-backends.md [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        ENV AIRFLOW__SECRETS__BACKEND="airflow.contrib.secrets.aws_systems_manager.SystemsManagerParameterStoreBackend"
+        ENV AIRFLOW__SECRETS__BACKEND_KWARGS="{"connections_prefix": "/airflow/connections", "variables_prefix": "/airflow/variables"}"
+
+   [Note that you can also add `profile` param to your kwargs](https://airflow.apache.org/docs/1.10.10/howto/use-alternative-secrets-backend.html) if you would like to reference an AWS profile instead of connecting via env vars.
+
+   This tells Airflow to look for connection information at the `airflow/connections/*` path in your SSM instan ce.
+
+   You can read up on [the full list of available kwargs for this integration](https://airflow.readthedocs.io/en/latest/_api/airflow/providers/amazon/aws/secrets/systems_manager/index.html) if you would like to further customize what the interaction between Airflow and SSM server will look like.
+
+### Test Your Connection
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:160]: Might be a me/noob thing, but is this process the same one you outlined earlier for Hashicorp Vault? Get the feeling more explanation on what a user needs to do here would be helpful

File: v0.15/secrets-backends.md [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+Test the connection to AWS SSM Parameter Store by calling the `get_conn_uri` method and passing it the conn_id of the secret you uploaded to your parameter store. If you followed our example, this would be `smtp_default`.
+
+### Deploy to Astronomer
+
+Once you've confirmed that your connections are being applied correctly locally, head to the Astronomer UI to set your production environment's environment variables. In the `Deployment>Configure>Environment Variables` section of our UI, set your `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.
+
+Once your env vars are set, deploy your local project up to Astronomer via `astro deploy`. You now should be able to observe your connection information being pulled from AWS SSM Parameter Store on Astronomer!
```
Comment (paolaperaza) [inline v0.15/secrets-backends.md:168]: `env vars` > `Environment Variables`

File: v0.15/secrets-backends.md (hunk 1 above)
Comment (vparekh94) [reply v0.15/secrets-backends.md:8]: Hm @paolaperaza I see what you are saying, but do we need to spend words "bad mouthing" airflow there (for lack of better words). 

Airflow 1.10.10 introduced the ability for users to sync secrets across.... - making it much easier to securely and automatically handle connections and variables.

File: v0.15/secrets-backends.md [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+## Hashicorp Vault
+
+In this section, we'll walk through how to use Vault as a secrets backend for your Airflow environment running on Astronomer.
```
[the remark this answers is not in the document]
Comment (vparekh94) [reply v0.15/secrets-backends.md:12]: :+1: on that we should standardize Airflow Deployment in internal docs.

File: v0.15/secrets-backends.md (hunk 1 above)
Comment (petedejoy) [reply v0.15/secrets-backends.md]: I agree- don't need the verbose intro, especially for a technical doc. Shortened it up.

File: v0.15/secrets-backends.md [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+## Hashicorp Vault
```
[the remark this answers is not in the document]
Comment (petedejoy) [reply v0.15/secrets-backends.md:10]: Done (TIL prerequisites is one word)

File: v0.15/secrets-backends.md [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+### Deploy a Vault Server
+
+To start, you will need to have a vault server running. If you do not have one deployed and would like to test this feature, [this Heroku Element](https://elements.heroku.com/buttons/pallavkothari/vault) for deploying a lightweight Vault server to Heroku is a good starting point for learning the ropes and getting your hands dirty with the Vault CLI. Deploying a local server via the instructions in [our Airflow and Vault guide](https://astronomer.io/guides/airflow-and-hashicorp-vault) should also work for testing.
+
+Once your server is deployed, be sure to save your `Root Token` and the URL of your vault server (if you deployed a local server, this should be `http://127.0.0.1:8200/`) for use in Airflow.
+  
+### Write a Connection to Vault
+
+Next, you'll need to get a connection uri into your Vault server that you can then use in your Airflow environment. We recommend setting the path to `secret/connections/<your-connection>` so that we can keep all of our Airflow connections organized in the `connections` directory of the mount point.
```
[the remark this answers is not in the document]
Comment (petedejoy) [reply v0.15/secrets-backends.md]: Hmm, do we think that creating a conn uri is scoped to this doc? That's something that's pretty core to Airflow/general connection management. I'd say that if a user doesn't know they need to do that/how to do that, this probably isn't the right doc to be starting in.

File: v0.15/secrets-backends.md (hunk 14 above)
Comment (petedejoy) [reply v0.15/secrets-backends.md:177]: Not quite and this is admittedly a bit weak because I don't have any experience with SSM and don't really know how to go lower-level than to link out to the appropriate docs 🤓

**Comment by petedejoy** [comment]
Thanks @paolaperaza! I don't _think_ they're not redundant- configuring the dockerfile is telling the image to look for specific env vars once it's built (ie. either pull those from a `.env` locally or from the env vars injected through the Astro UI. If you don't do that, the image won't be built with those.

@kaxil can you confirm?

**Comment by paolaperaza** [comment]
@petedejoy Aren't these two sections saying the same thing though (i.e. set "AWS Access Key ID" and "AWS Secret Access Key" in your `.env` file)?

> ### Connect to Parameter Store
> 
> Start by establishing a connection to your AWS Parameter store in your Airflow project. To do so, set the following Environment Variables in your Airflow project's `.env` file:
> 
> ```
> AWS_ACCESS_KEY_ID="<YOUR-AWS-ACCESS-KEY>"
> AWS_SECRET_ACCESS_KEY="<YOUR-AWS-SECRET-ACCESS-KEY>"
> ```
> 
> This will allow you to test this feature locally. From there, we'll input similar Environment Variables on a running Airflow Deployment on Astronomer.
> 
> ### Configure Your Dockerfile
> 
> Now that you have a connection established in AWS, configure your `Dockerfile` on Astronomer to use this AWS SSM Parameter Store instance server as the default secrets backend for your local Airflow instance.
> 
> To set this configuration up, follow the steps below.
> 
> 1. Add the following environment variables to the `.env` file in your project directory to be used in your _local_ Airflow instance:
> 
>         AWS_ACCESS_KEY_ID="YOUR-AWS-KEY"
>         AWS_SECRET_ACCESS_KEY="YOUR-AWS-SECRET-KEY"

**Comment by petedejoy** [comment]
Yeah, I'm just not sure if we even need to add those to our dockerfile or if just setting them as env vars will do the trick- will need to wait for Kaxil's advice.

File: v0.15/secrets-backends.md [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+### Configure Your Dockerfile
+
+Now that you have a connection established in AWS, configure your `Dockerfile` on Astronomer to use this AWS SSM Parameter Store instance server as the default secrets backend for your local Airflow instance.
+
+To set this configuration up, follow the steps below.
+
+1. Add the following environment variables to the `.env` file in your project directory to be used in your _local_ Airflow instance:
+
+        AWS_ACCESS_KEY_ID="YOUR-AWS-KEY"
+        AWS_SECRET_ACCESS_KEY="YOUR-AWS-SECRET-KEY"
```
Comment (kaxil) [inline v0.15/secrets-backends.md:200]: Yeah this section is definitely a duplicate of L181 to 189

**Comment by kaxil** [comment]
There were 2 comments to the PR that I thought I had posted a long time but I forgot to press "Submit Review" button

File: v0.15/secrets-backends.md (hunk 19 above)
Comment (paolaperaza) [reply v0.15/secrets-backends.md:212]: @kaxil Got it, ok. So which section is necessary? Should we take out the whole "Connect to Parameter Store" section and leave everything in "Configure your Dockerfile"?

**Comment by paolaperaza** [comment]
@kaxil You think we should leave out SSM Parameter store entirely from this doc? I'll defer to you either way - if what we have on there isn't complete but is accurate, I think I'm ok with that too.

Let me know what you think re: leaving the "Test your Connection" section and I'll adjust and we can hopefully merge tomorrow.

One last related question: In the Vault "Test your Connection Locally" section, we do ask users to input their Vault Root Token and Vault URL as a part of their `.env` file, but also as part of their Dockerfile. Do we need both?

File: v0.15/secrets-backends.md (hunk 19 above)
Comment (kaxil) [reply v0.15/secrets-backends.md:212]: Yeah, we can remove the whole of `"Connect to Parameter Store"`

**Comment by kaxil** [comment]
> @kaxil You think we should leave out SSM Parameter store entirely from this doc? I'll defer to you either way - if what we have on there isn't complete but is accurate, I think I'm ok with that too.
> 
> Let me know what you think re: leaving the "Test your Connection" section and I'll adjust and we can hopefully merge tomorrow.
> 
> One last related question: In the Vault "Test your Connection Locally" section, we do ask users to input their Vault Root Token and Vault URL as a part of their `.env` file, but also as part of their Dockerfile. Do we need both?

It is accurate so we can keep it there for now. And add more details later on when we test it with the environment

File: v0.15/secrets-backends.md [hunk 20]
```diff
@@ -0,0 +1,185 @@
+---
+title: "Managing Secrets"
+description: "Options for secrets management backends on Astronomer."
+date: 2019-05-28T00:00:00.000Z
+slug: "secrets-backends"
+---
+
+As of Airflow 1.10.10, users can [manage connections from an external secrets backend](https://airflow.apache.org/docs/1.10.10/howto/use-alternative-secrets-backend.html) such as [Hashicorp Vault](https://www.vaultproject.io/), [AWS SSM](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html), or [GCP Secret Manager](https://www.google.com/search?q=gcp+secrets+manager&oq=gcp+secrets+manager&aqs=chrome.0.0l8.2480j0j4&sourceid=chrome&ie=UTF-8).
```
[the remark this answers is not in the document]
Comment (paolaperaza) [reply v0.15/secrets-backends.md]: Applied this change! Thanks @kaxil

**Comment by paolaperaza** [comment]
Alright, I think we are finally _set_ here! Merging. Thanks a ton for the last review @kaxil - removed the "Connect to Parameter Store" section but kept the rest.
